Do we want to fix these gcc warnings?

## -Wswitch-default

```
src/ciphers/multi2.c:96:7: error: switch missing default case [-Werror=switch-default]
       switch (n<=4 ? n : ((n-1)%4)+1) {
       ^~~~~~
src/ciphers/twofish/twofish.c:247:3: error: switch missing default case [-Werror=switch-default]
   switch (k) {
   ^~~~~~
src/hashes/tiger.c:569:5: error: switch missing default case [-Werror=switch-default]
     switch (mul) {
     ^~~~~~
src/pk/asn1/der/choice/der_decode_choice.c:51:8: error: switch missing default case [-Werror=switch-default]
        switch (list[x].type) {
        ^~~~~~
src/pk/asn1/der/custom_type/der_decode_custom_type.c:163:8: error: switch missing default case [-Werror=switch-default]
        switch (type) {
        ^~~~~~
src/pk/asn1/der/custom_type/der_encode_custom_type.c:93:8: error: switch missing default case [-Werror=switch-default]
        switch (type) {
        ^~~~~~
src/pk/asn1/der/custom_type/der_length_custom_type.c:69:8: error: switch missing default case [-Werror=switch-default]
        switch (type) {
        ^~~~~~
src/pk/asn1/der/sequence/der_decode_sequence_multi.c:52:8: error: switch missing default case [-Werror=switch-default]
        switch (type) {
        ^~~~~~
src/pk/asn1/der/sequence/der_decode_sequence_multi.c:101:8: error: switch missing default case [-Werror=switch-default]
        switch (type) {
        ^~~~~~
```

## -Wswitch-enum

```
src/pk/asn1/der/sequence/der_sequence_free.c: In function ‘der_sequence_free’:
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_EOL’ not handled in switch [-Werror=switch-enum]
       switch (in->type) {
       ^~~~~~
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_BOOLEAN’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_SHORT_INTEGER’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_BIT_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_OCTET_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_NULL’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_OBJECT_IDENTIFIER’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_IA5_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_PRINTABLE_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_UTF8_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_UTCTIME’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_CHOICE’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_SEQUENCE’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_SET’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_RAW_BIT_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_TELETEX_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_GENERALIZEDTIME’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_free.c:46:7: error: enumeration value ‘LTC_ASN1_CUSTOM_TYPE’ not handled in switch [-Werror=switch-enum]

src/pk/asn1/der/sequence/der_sequence_shrink.c: In function ‘der_sequence_shrink’:
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_EOL’ not handled in switch [-Werror=switch-enum]
       switch (in->type) {
       ^~~~~~
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_BOOLEAN’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_INTEGER’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_SHORT_INTEGER’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_BIT_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_OCTET_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_NULL’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_OBJECT_IDENTIFIER’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_IA5_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_PRINTABLE_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_UTF8_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_UTCTIME’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_CHOICE’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_SETOF’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_RAW_BIT_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_TELETEX_STRING’ not handled in switch [-Werror=switch-enum]
src/pk/asn1/der/sequence/der_sequence_shrink.c:34:7: error: enumeration value ‘LTC_ASN1_GENERALIZEDTIME’ not handled in switch [-Werror=switch-enum]
```

## -Wmissing-noreturn

UPDATE: fixed by #428

## -Wmissing-prototypes (some functions should be decared as static)

UPDATE: fixed by #428

## -Wmissing-declarations (similar to previous one)

UPDATE: fixed by #428

--- a/demos/ltcrypt.c
+++ b/demos/ltcrypt.c
@@ -18,7 +18,7 @@
 
 #include <tomcrypt.h>
 
-int usage(char *name)
+static int NORETURN usage(char *name)
 {
    int x;
 

--- a/demos/tv_gen.c
+++ b/demos/tv_gen.c
@@ -176,9 +176,9 @@ static void hmac_gen(void)
    fclose(out);
 }
 
+#ifdef LTC_OMAC
 static void omac_gen(void)
 {
-#ifdef LTC_OMAC
    unsigned char key[MAXBLOCKSIZE], output[MAXBLOCKSIZE], input[MAXBLOCKSIZE*2+2];
    int err, x, y, z, kl;
    FILE *out;
@@ -234,12 +234,12 @@ static void omac_gen(void)
       fprintf(out, "\n");
    }
    fclose(out);
-#endif
 }
+#endif
 
+#ifdef LTC_PMAC
 static void pmac_gen(void)
 {
-#ifdef LTC_PMAC
    unsigned char key[MAXBLOCKSIZE], output[MAXBLOCKSIZE], input[MAXBLOCKSIZE*2+2];
    int err, x, y, z, kl;
    FILE *out;
@@ -295,12 +295,12 @@ static void pmac_gen(void)
       fprintf(out, "\n");
    }
    fclose(out);
-#endif
 }
+#endif
 
+#ifdef LTC_EAX_MODE
 static void eax_gen(void)
 {
-#ifdef LTC_EAX_MODE
    int err, kl, x, y1, z;
    FILE *out;
    unsigned char key[MAXBLOCKSIZE], nonce[MAXBLOCKSIZE*2], header[MAXBLOCKSIZE*2],
@@ -361,12 +361,12 @@ static void eax_gen(void)
       fprintf(out, "\n");
    }
    fclose(out);
-#endif
 }
+#endif
 
+#ifdef LTC_OCB_MODE
 static void ocb_gen(void)
 {
-#ifdef LTC_OCB_MODE
    int err, kl, x, y1, z;
    FILE *out;
    unsigned char key[MAXBLOCKSIZE], nonce[MAXBLOCKSIZE*2],
@@ -430,12 +430,12 @@ static void ocb_gen(void)
       fprintf(out, "\n");
    }
    fclose(out);
-#endif
 }
+#endif
 
+#ifdef LTC_OCB3_MODE
 static void ocb3_gen(void)
 {
-#ifdef LTC_OCB3_MODE
    int err, kl, x, y1, z, noncelen;
    FILE *out;
    unsigned char key[MAXBLOCKSIZE], nonce[MAXBLOCKSIZE*2],
@@ -500,12 +500,12 @@ static void ocb3_gen(void)
       fprintf(out, "\n");
    }
    fclose(out);
-#endif
 }
+#endif
 
+#ifdef LTC_CCM_MODE
 static void ccm_gen(void)
 {
-#ifdef LTC_CCM_MODE
    int err, kl, x, y1, z;
    FILE *out;
    unsigned char key[MAXBLOCKSIZE], nonce[MAXBLOCKSIZE*2],
@@ -569,12 +569,12 @@ static void ccm_gen(void)
       fprintf(out, "\n");
    }
    fclose(out);
-#endif
 }
+#endif
 
+#ifdef LTC_GCM_MODE
 static void gcm_gen(void)
 {
-#ifdef LTC_GCM_MODE
    int err, kl, x, y1, z;
    FILE *out;
    unsigned char key[MAXBLOCKSIZE], plaintext[MAXBLOCKSIZE*2], tag[MAXBLOCKSIZE];
@@ -632,8 +632,8 @@ static void gcm_gen(void)
       fprintf(out, "\n");
    }
    fclose(out);
-#endif
 }
+#endif
 
 static void base64_gen(void)
 {
@@ -701,9 +701,9 @@ static void ecc_gen(void)
    fclose(out);
 }
 
+#ifdef LTC_LRW_MODE
 static void lrw_gen(void)
 {
-#ifdef LTC_LRW_MODE
    FILE *out;
    unsigned char tweak[16], key[16], iv[16], buf[1024];
    int x, y, err;
@@ -765,8 +765,8 @@ static void lrw_gen(void)
        lrw_done(&lrw);
    }
    fclose(out);
-#endif
 }
+#endif
 
 int main(void)
 {

--- a/demos/tv_gen.c
+++ b/demos/tv_gen.c
@@ -8,7 +8,7 @@
  */
 #include "tomcrypt_private.h"
 
-void hash_gen(void)
+static void hash_gen(void)
 {
    unsigned char md[MAXBLOCKSIZE], *buf;
    unsigned long outlen, x, y, z;
@@ -49,7 +49,7 @@ void hash_gen(void)
    fclose(out);
 }
 
-void cipher_gen(void)
+static void cipher_gen(void)
 {
    unsigned char *key, pt[MAXBLOCKSIZE];
    unsigned long x, y, z, w;
@@ -124,7 +124,7 @@ void cipher_gen(void)
   fclose(out);
 }
 
-void hmac_gen(void)
+static void hmac_gen(void)
 {
    unsigned char key[MAXBLOCKSIZE], output[MAXBLOCKSIZE], *input;
    int x, y, z, err;
@@ -176,7 +176,7 @@ void hmac_gen(void)
    fclose(out);
 }
 
-void omac_gen(void)
+static void omac_gen(void)
 {
 #ifdef LTC_OMAC
    unsigned char key[MAXBLOCKSIZE], output[MAXBLOCKSIZE], input[MAXBLOCKSIZE*2+2];
@@ -237,7 +237,7 @@ void omac_gen(void)
 #endif
 }
 
-void pmac_gen(void)
+static void pmac_gen(void)
 {
 #ifdef LTC_PMAC
    unsigned char key[MAXBLOCKSIZE], output[MAXBLOCKSIZE], input[MAXBLOCKSIZE*2+2];
@@ -298,7 +298,7 @@ void pmac_gen(void)
 #endif
 }
 
-void eax_gen(void)
+static void eax_gen(void)
 {
 #ifdef LTC_EAX_MODE
    int err, kl, x, y1, z;
@@ -364,7 +364,7 @@ void eax_gen(void)
 #endif
 }
 
-void ocb_gen(void)
+static void ocb_gen(void)
 {
 #ifdef LTC_OCB_MODE
    int err, kl, x, y1, z;
@@ -433,7 +433,7 @@ void ocb_gen(void)
 #endif
 }
 
-void ocb3_gen(void)
+static void ocb3_gen(void)
 {
 #ifdef LTC_OCB3_MODE
    int err, kl, x, y1, z, noncelen;
@@ -503,7 +503,7 @@ void ocb3_gen(void)
 #endif
 }
 
-void ccm_gen(void)
+static void ccm_gen(void)
 {
 #ifdef LTC_CCM_MODE
    int err, kl, x, y1, z;
@@ -572,7 +572,7 @@ void ccm_gen(void)
 #endif
 }
 
-void gcm_gen(void)
+static void gcm_gen(void)
 {
 #ifdef LTC_GCM_MODE
    int err, kl, x, y1, z;
@@ -635,7 +635,7 @@ void gcm_gen(void)
 #endif
 }
 
-void base64_gen(void)
+static void base64_gen(void)
 {
    FILE *out;
    unsigned char src[32], ch;
@@ -655,11 +655,11 @@ void base64_gen(void)
    fclose(out);
 }
 
-void math_gen(void)
+static void math_gen(void)
 {
 }
 
-void ecc_gen(void)
+static void ecc_gen(void)
 {
    FILE         *out;
    unsigned char str[512];
@@ -701,7 +701,7 @@ void ecc_gen(void)
    fclose(out);
 }
 
-void lrw_gen(void)
+static void lrw_gen(void)
 {
 #ifdef LTC_LRW_MODE
    FILE *out;

--- a/tests/der_test.c
+++ b/tests/der_test.c
@@ -1301,7 +1301,7 @@ static void der_Xcode_test(void)
 }
 
 #ifdef LTC_TEST_READDIR
-int _der_decode_sequence_flexi(const void *in, unsigned long inlen, void* ctx)
+static int _der_decode_sequence_flexi(const void *in, unsigned long inlen, void* ctx)
 {
    ltc_asn1_list** list = ctx;
    if (der_decode_sequence_flexi(in, &inlen, list) == CRYPT_OK) {

--- a/tests/ecc_test.c
+++ b/tests/ecc_test.c
@@ -209,7 +209,7 @@ done:
    return err;
 }
 
-int _ecc_old_api(void)
+static int _ecc_old_api(void)
 {
    unsigned char buf[4][4096], ch;
    unsigned long x, y, z, s;
@@ -350,7 +350,7 @@ int _ecc_old_api(void)
    return CRYPT_OK;
 }
 
-int _ecc_new_api(void)
+static int _ecc_new_api(void)
 {
    const char* names[] = {
 #ifdef LTC_ECC_SECP112R1
@@ -517,7 +517,7 @@ int _ecc_new_api(void)
    return CRYPT_OK;
 }
 
-int _ecc_key_cmp(const int should_type, const ecc_key *should, const ecc_key *is)
+static int _ecc_key_cmp(const int should_type, const ecc_key *should, const ecc_key *is)
 {
    if (should_type != is->type)                               return CRYPT_ERROR;
    if (should_type == PK_PRIVATE) {
@@ -536,7 +536,7 @@ int _ecc_key_cmp(const int should_type, const ecc_key *should, const ecc_key *is
    return CRYPT_OK;
 }
 
-int _ecc_import_export(void) {
+static int _ecc_import_export(void) {
    const ltc_ecc_curve *cu;
    ecc_key key, pri, pub;
    unsigned char out[300];

--- a/tests/no_prng.c
+++ b/tests/no_prng.c
@@ -6,7 +6,9 @@
  * The library is free for all purposes without any express
  * guarantee it works.
  */
+
 #include "tomcrypt.h"
+#include "tomcrypt_test.h"
 
 /**
   @file no_prng.c
@@ -29,7 +31,7 @@ typedef struct
   @param prng     [out] The PRNG state to initialize
   @return CRYPT_OK if successful
 */
-int no_prng_start(prng_state *prng)
+static int no_prng_start(prng_state *prng)
 {
    no_prng_desc_t *no_prng = (no_prng_desc_t*) prng;
    LTC_ARGCHK(no_prng != NULL);
@@ -47,7 +49,7 @@ int no_prng_start(prng_state *prng)
   @param prng     PRNG state to update
   @return CRYPT_OK if successful
 */
-int no_prng_add_entropy(const unsigned char *in, unsigned long inlen, prng_state *prng)
+static int no_prng_add_entropy(const unsigned char *in, unsigned long inlen, prng_state *prng)
 {
    no_prng_desc_t *no_prng = (no_prng_desc_t*) prng;
    LTC_ARGCHK(no_prng != NULL);
@@ -68,7 +70,7 @@ int no_prng_add_entropy(const unsigned char *in, unsigned long inlen, prng_state
   @param prng   The PRNG to make active
   @return CRYPT_OK if successful
 */
-int no_prng_ready(prng_state *prng)
+static int no_prng_ready(prng_state *prng)
 {
     LTC_ARGCHK(prng != NULL);
 
@@ -82,7 +84,7 @@ int no_prng_ready(prng_state *prng)
   @param prng     The active PRNG to read from
   @return Number of octets read
 */
-unsigned long no_prng_read(unsigned char *out, unsigned long outlen, prng_state *prng)
+static unsigned long no_prng_read(unsigned char *out, unsigned long outlen, prng_state *prng)
 {
    no_prng_desc_t *no_prng = (no_prng_desc_t*) prng;
    LTC_ARGCHK(no_prng != NULL);
@@ -101,7 +103,7 @@ unsigned long no_prng_read(unsigned char *out, unsigned long outlen, prng_state 
   @param prng   The PRNG to terminate
   @return CRYPT_OK if successful
 */
-int no_prng_done(prng_state *prng)
+static int no_prng_done(prng_state *prng)
 {
    LTC_UNUSED_PARAM(prng);
    return CRYPT_OK;
@@ -114,7 +116,7 @@ int no_prng_done(prng_state *prng)
   @param prng      The PRNG to export
   @return CRYPT_OK if successful
 */
-int no_prng_export(unsigned char *out, unsigned long *outlen, prng_state *prng)
+static int no_prng_export(unsigned char *out, unsigned long *outlen, prng_state *prng)
 {
    LTC_UNUSED_PARAM(out);
    LTC_UNUSED_PARAM(outlen);
@@ -129,7 +131,7 @@ int no_prng_export(unsigned char *out, unsigned long *outlen, prng_state *prng)
   @param prng     The PRNG to import
   @return CRYPT_OK if successful
 */
-int no_prng_import(const unsigned char *in, unsigned long inlen, prng_state *prng)
+static int no_prng_import(const unsigned char *in, unsigned long inlen, prng_state *prng)
 {
    LTC_UNUSED_PARAM(in);
    LTC_UNUSED_PARAM(inlen);
@@ -141,7 +143,7 @@ int no_prng_import(const unsigned char *in, unsigned long inlen, prng_state *prn
   PRNG self-test
   @return CRYPT_OK if successful, CRYPT_NOP if self-testing has been disabled
 */
-int no_prng_test(void)
+static int no_prng_test(void)
 {
    return CRYPT_OK;
 }

